File: src/Transpiler.Backends/Runtime/managed.mjs
// Managed values copy; managed references alias storage. Neither is a plain host object convention.
class CliValue extends CliObject {}
class ManagedRuntime extends CliRuntime {
    constructor(metadata, write = null) {
        super(metadata, write);
        this.parents['System.ArgumentNullException'] = 'System.ArgumentException';
        this.parents['System.OperationCanceledException'] = 'System.SystemException';
        this.parents['System.Threading.Tasks.TaskCanceledException'] = 'System.OperationCanceledException';
    }
    default(type) {
        const descriptor = this.meta.types[type];
        if (descriptor?.enumType) return super.default(descriptor.enumType);
        if (descriptor?.valueType) {
            const value = new CliValue(type);
            for (const key of descriptor.fields) {
                const field = this.meta.fields[key];
                if (!field.static) value.fields[key] = this.default(field.type);
            }
            return value;
        }
        return super.default(type);
    }
    copy_value(value) {
        if (!(value instanceof CliValue)) return value;
        const copy = new CliValue(value.type);
        for (const [key, item] of Object.entries(value.fields)) copy.fields[key] = this.copy_value(item);
        return copy;
    }
    coerce(value, type) {
        const descriptor = this.meta.types[type];
        if (descriptor?.enumType) return super.coerce(value, descriptor.enumType);
        if (descriptor?.valueType) {
            if (!(value instanceof CliValue) || value.type !== type) throw new TypeError('Invalid managed value representation for ' + type);
            return this.copy_value(value);
        }
        return super.coerce(value, type);
    }
    dereference(value) { return value instanceof CliRef ? value.get() : value; }
    format(value, type = 'System.Object') {
        value = this.dereference(value);
        if (value instanceof CliBox) return this.format(value.value, value.type);
        const descriptor = this.meta.types[type];
        if (descriptor?.enumType) return this.enum_text(value, descriptor);
        if (value instanceof CliObject) {
            let name = value.type;
            if (name.startsWith('[')) name = name.slice(name.indexOf(']') + 1);
            return name.replaceAll('<', '[').replaceAll('>', ']');
        }
        return super.format(value, type);
    }
    enum_text(value, descriptor) {
        const width = ['System.Int64', 'System.UInt64'].includes(descriptor.enumType) ? 64 : 32;
        const bits = BigInt.asUintN(width, BigInt(value));
        const entries = descriptor.enumValues.map(item => [BigInt.asUintN(width, BigInt(item.value)), item.name]);
        entries.sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0);
        for (const [number, name] of entries) if (number === bits) return name;
        if (descriptor.enumFlags && bits !== 0n) {
            let remaining = bits; const names = [];
            for (let i = entries.length - 1; i >= 0; --i) {
                const [number, name] = entries[i];
                if (number !== 0n && (remaining & number) === number) { remaining &= ~number; names.unshift(name); }
            }
            if (remaining === 0n) return names.join(', ');
        }
        return String(value);
    }
    field_get(obj, key) { return this.coerce(super.field_get(this.dereference(obj), key), this.meta.fields[key].type); }
    field_set(obj, key, value) { super.field_set(this.dereference(obj), key, value); }
    field_ref(obj, key) {
        const field = this.meta.fields[key];
        if (field.static) { this.ensure(field.owner); return this.cell_ref(this.statics, key, field.type); }
        this.nonnull(this.dereference(obj));
        // Keep the location, not a stale object captured before its enclosing struct is replaced.
        return new CliRef(() => this.nonnull(this.dereference(obj)).fields[key], value => this.field_set(obj, key, value));
    }
    function_pointer(methodId, receiver, virtual = false) {
        const method = this.meta.methods[methodId];
        if (virtual) {
            this.nonnull(receiver);
            if (method.virtual) methodId = this.meta.types[receiver.type]?.vtable?.[method.slot] ?? methodId;
        }
        return new CliFunction(methodId);
    }
    new_object(methodId, args) {
        const method = this.meta.methods[methodId], descriptor = this.meta.types[method.type];
        if (method.intrinsic === 'delegate.ctor') {
            if (!(args[1] instanceof CliFunction)) throw new TypeError('Delegate construction requires a translated method pointer');
            return new CliDelegate(method.type, [[args[0], args[1].method]]);
        }
        if (descriptor?.valueType) {
            this.ensure(method.type);
            const cell = [this.default(method.type)];
            this.call(methodId, [this.cell_ref(cell, 0, method.type), ...args]);
            return cell[0];
        }
        return super.new_object(methodId, args);
    }
    assignable(actual, target) {
        if (actual === target) return true;
        if (actual.endsWith('[]') && target.endsWith('[]')) {
            const a = actual.slice(0, -2), b = target.slice(0, -2);
            return this.default(a) === null && this.default(b) === null && this.assignable(a, b);
        }
        const todo = [actual], seen = new Set();
        while (todo.length) {
            const current = todo.pop();
            if (current === target) return true;
            if (current == null || seen.has(current)) continue;
            seen.add(current);
            const descriptor = this.meta.types[current];
            const expected = this.meta.types[target];
            if (descriptor && expected && descriptor.definition === expected.definition && descriptor.arguments.length && descriptor.variance.length === descriptor.arguments.length) {
                const reference = type => this.default(type) === null;
                if (descriptor.arguments.every((a, i) => {
                    const b = expected.arguments[i], v = descriptor.variance[i];
                    return a === b || reference(a) && reference(b) && (v === 1 ? this.assignable(a, b) : v === 2 && this.assignable(b, a));
                })) return true;
            }
            if (descriptor) todo.push(...descriptor.interfaces);
            todo.push(this.parents[current]);
        }
        return false;
    }
    enumerable_element(type) {
        const descriptor = this.meta.types[type];
        return descriptor?.definition === 'System.Collections.Generic.IEnumerable`1' ? descriptor.arguments[0] : null;
    }
    is_type(value, target) {
        if (value instanceof CliArray || value instanceof CliString) {
            if (target === 'System.Collections.IEnumerable') return true;
            const wanted = this.enumerable_element(target), actual = value instanceof CliArray ? value.element : 'System.Char';
            if (wanted !== null) return actual === wanted || this.default(actual) === null && this.default(wanted) === null && this.assignable(actual, wanted);
        }
        if (value instanceof CliDelegate) return ['System.Object', 'System.Delegate', 'System.MulticastDelegate', value.type].includes(target);
        if (value instanceof CliBox) return target === 'System.Object' || target === 'System.ValueType' || this.assignable(value.type, target);
        return super.is_type(value, target);
    }
    box(value, type) { return this.default(type) !== null ? new CliBox(type, this.coerce(value, type)) : value; }
    unbox(value, type, address = false) {
        if (this.default(type) === null && !address) return this.cast(value, type);
        this.nonnull(value);
        if (!(value instanceof CliBox) || value.type !== type) this.fail('System.InvalidCastException', 'Specified cast is not valid.');
        return address ? new CliRef(() => value.value, v => { value.value = this.coerce(v, type); }) : this.coerce(value.value, type);
    }
    call(methodId, args, virtual = false, constrained = null) {
        let method = this.meta.methods[methodId];
        if (constrained !== null && method.intrinsic === 'object.string') return this.string(this.format(this.nonnull(this.dereference(args[0])), constrained));
        if (virtual) {
            const original = this.nonnull(args[0]); let receiver = this.nonnull(this.dereference(original));
            const wanted = this.enumerable_element(method.type);
            if ((receiver instanceof CliArray || receiver instanceof CliString) && method.name === 'GetEnumerator' && (wanted !== null || method.type === 'System.Collections.IEnumerable')) {
                if (receiver instanceof CliString) receiver = new CliArray('System.Char', Array.from({length: receiver.text.length}, (_, i) => receiver.text.charCodeAt(i)));
                const constructor = this.meta.arrayEnumerators[(wanted ?? receiver.element) + '[]'];
                if (!constructor) throw new Error('Array enumeration contract was not linked');
                return this.new_object(constructor, [receiver]);
            }
            const actualType = receiver instanceof CliObject || receiver instanceof CliBox ? receiver.type : null;
            if (method.virtual && actualType) {
                methodId = this.meta.types[actualType]?.vtable?.[method.slot] ?? methodId;
                method = this.meta.methods[methodId];
            }
            const descriptor = this.meta.types[method.type];
            if (descriptor?.valueType) {
                if (receiver instanceof CliBox) args = [this.unbox(receiver, receiver.type, true), ...args.slice(1)];
                else if (original instanceof CliRef) args = [original, ...args.slice(1)];
                else throw new TypeError('Value-type method requires a managed address');
            } else args = [receiver, ...args.slice(1)];
        }
        return super.call(methodId, args, false);
    }
    async await_export(name, args, options = {}) {
        const maxSteps = options.maxSteps ?? 100000;
        if (!Number.isSafeInteger(maxSteps) || maxSteps < 1) throw new TypeError('maxSteps must be a positive integer');
        const yieldHost = options.yieldHost ?? (() => new Promise(resolve => setTimeout(resolve, 0)));
        if (typeof yieldHost !== 'function') throw new TypeError('yieldHost must be a function');
        const candidates = Object.entries(this.meta.exports).filter(([k]) => k === name || k.split('(')[0] === name);
        if (candidates.length !== 1) throw new Error('Use an unambiguous exported method signature');
        const binding = this.meta.asyncBindings[this.meta.methods[candidates[0][1]].returns];
        const value = this.invoke_export(name, args);
        if (!binding) return value;
        for (let step = 0; step < maxSteps; ++step) {
            options.signal?.throwIfAborted();
            if (this.call(binding.completed, [value])) {
                const cell = [this.call(binding.getAwaiter, [value])];
                const result = this.call(binding.getResult, [this.cell_ref(cell, 0, binding.awaiterType)]);
                if (result instanceof CliString) return result.text;
                return binding.resultType === 'System.Boolean' ? Boolean(result) : result;
            }
            this.call(binding.pump, []);
            await yieldHost();
        }
        throw new Error('Cooperative task exceeded the host pump step budget');
    }
    delegate_equal(a, b) {
        if (a === b) return true;
        if (a === null || b === null || a.type !== b.type || a.invocations.length !== b.invocations.length) return false;
        return a.invocations.every((x, i) => x[0] === b.invocations[i][0] && x[1] === b.invocations[i][1]);
    }
    external(method, args) {
        const op = method.intrinsic;
        if (op === 'environment.thread') return 1;
        if (op === 'delegate.invoke') {
            const delegate = this.nonnull(args[0]); let result = null;
            for (const [target, pointer] of delegate.invocations) {
                const callee = this.meta.methods[pointer]; let receiver = target;
                if (callee.instance && this.meta.types[callee.type]?.valueType) receiver = this.unbox(target, callee.type, true);
                result = this.call(pointer, [...(callee.instance ? [receiver] : []), ...args.slice(1)]);
            }
            return result;
        }
        if (op === 'delegate.equals' || op === 'delegate.not-equals') {
            const result = this.delegate_equal(args[0], args[1]); return Number(op === 'delegate.not-equals' ? !result : result);
        }
        if (op === 'delegate.combine' || op === 'delegate.remove') {
            const [a, b] = args;
            if (a === null) return op === 'delegate.combine' ? b : null;
            if (b === null) return a;
            if (a.type !== b.type) this.fail('System.ArgumentException', 'Delegates must be of the same type.');
            if (op === 'delegate.combine') return new CliDelegate(a.type, [...a.invocations, ...b.invocations]);
            for (let start = a.invocations.length - b.invocations.length; start >= 0; --start) {
                const part = new CliDelegate(a.type, a.invocations.slice(start, start + b.invocations.length));
                if (this.delegate_equal(part, b)) {
                    const remaining = [...a.invocations.slice(0, start), ...a.invocations.slice(start + b.invocations.length)];
                    return remaining.length ? new CliDelegate(a.type, remaining) : null;
                }
            }
            return a;
        }
        return super.external(method, args);
    }

}

class CliFunction { constructor(method) { this.method = method; } }
class CliDelegate { constructor(type, invocations) { this.type = type; this.invocations = invocations; } }

File: src/Transpiler.Backends/Runtime/managed.py
# Managed storage semantics shared by ordinary code and translated C# library implementations.
class CliValue(CliObject):
    __slots__ = ()


class ManagedRuntime(CliRuntime):
    def __init__(self, metadata, write=None):
        super().__init__(metadata, write)
        self.parents['System.ArgumentNullException'] = 'System.ArgumentException'
        self.parents['System.OperationCanceledException'] = 'System.SystemException'
        self.parents['System.Threading.Tasks.TaskCanceledException'] = 'System.OperationCanceledException'

    def default(self, type_name):
        descriptor = self.meta['types'].get(type_name)
        if descriptor and descriptor['enumType']:
            return super().default(descriptor['enumType'])
        if descriptor and descriptor['valueType']:
            value = CliValue(type_name)
            for key in descriptor['fields']:
                field = self.meta['fields'][key]
                if not field['static']: value.fields[key] = self.default(field['type'])
            return value
        return super().default(type_name)

    def copy_value(self, value):
        if not isinstance(value, CliValue): return value
        copy = CliValue(value.type)
        copy.fields = {key: self.copy_value(item) for key, item in value.fields.items()}
        return copy

    def coerce(self, value, type_name):
        descriptor = self.meta['types'].get(type_name)
        if descriptor and descriptor['enumType']:
            return super().coerce(value, descriptor['enumType'])
        if descriptor and descriptor['valueType']:
            if not isinstance(value, CliValue) or value.type != type_name:
                raise TypeError('Invalid managed value representation for ' + type_name)
            return self.copy_value(value)
        return super().coerce(value, type_name)

    @staticmethod
    def dereference(value):
        return value.get() if isinstance(value, CliRef) else value

    def format(self, value, type_name='System.Object'):
        value = self.dereference(value)
        if isinstance(value, CliBox): return self.format(value.value, value.type)
        descriptor = self.meta['types'].get(type_name)
        if descriptor and descriptor['enumType']:
            return self.enum_text(value, descriptor)
        if isinstance(value, CliObject):
            name = value.type
            if name.startswith('['): name = name[name.index(']') + 1:]
            return name.replace('<', '[').replace('>', ']')
        return super().format(value, type_name)

    @staticmethod
    def enum_text(value, descriptor):
        width = 64 if descriptor['enumType'] in ('System.Int64', 'System.UInt64') else 32
        bits = int(value) & ((1 << width) - 1)
        entries = [(int(item['value']) & ((1 << width) - 1), item['name']) for item in descriptor['enumValues']]
        entries.sort(key=lambda item: item[0])
        for number, name in entries:
            if number == bits: return name
        if descriptor['enumFlags'] and bits:
            remaining, names = bits, []
            for number, name in reversed(entries):
                if number and remaining & number == number:
                    remaining &= ~number
                    names.insert(0, name)
            if remaining == 0: return ', '.join(names)
        return str(value)

    def field_get(self, obj, key):
        value = super().field_get(self.dereference(obj), key)
        return self.coerce(value, self.meta['fields'][key]['type'])

    def field_set(self, obj, key, value):
        super().field_set(self.dereference(obj), key, value)

    def field_ref(self, obj, key):
        field = self.meta['fields'][key]
        if field['static']:
            self.ensure(field['owner'])
            return self.cell_ref(self.statics, key, field['type'])
        self.nonnull(self.dereference(obj))
        # Re-evaluate the parent address on every access: s = new S() must not detach ref s.Field.
        return CliRef(lambda: self.nonnull(self.dereference(obj)).fields[key],
                      lambda value: self.field_set(obj, key, value))

    def function_pointer(self, method_id, receiver, virtual=False):
        method = self.meta['methods'][method_id]
        if virtual:
            receiver = self.nonnull(receiver)
            if method['virtual']:
                method_id = self.meta['types'].get(receiver.type, {}).get('vtable', {}).get(method['slot'], method_id)
        return CliFunction(method_id)

    def new_object(self, method_id, args):
        method = self.meta['methods'][method_id]
        if method['intrinsic'] == 'delegate.ctor':
            if not isinstance(args[1], CliFunction): raise TypeError('Delegate construction requires a translated method pointer')
            return CliDelegate(method['type'], [(args[0], args[1].method)])
        descriptor = self.meta['types'].get(method['type'])
        if descriptor and descriptor['valueType']:
            self.ensure(method['type'])
            cell = [self.default(method['type'])]
            self.call(method_id, [self.cell_ref(cell, 0, method['type'])] + args)
            return cell[0]
        return super().new_object(method_id, args)

    def assignable(self, actual, target):
        if actual == target: return True
        if actual.endswith('[]') and target.endswith('[]'):
            a, b = actual[:-2], target[:-2]
            return self.default(a) is None and self.default(b) is None and self.assignable(a, b)
        todo, seen = [actual], set()
        while todo:
            current = todo.pop()
            if current == target: return True
            if current is None or current in seen: continue
            seen.add(current)
            descriptor = self.meta['types'].get(current)
            expected = self.meta['types'].get(target)
            if descriptor and expected and descriptor['definition'] == expected['definition'] and descriptor['arguments']:
                def reference(name): return self.default(name) is None
                def matches(a, b, variance):
                    return a == b or reference(a) and reference(b) and ((variance == 1 and self.assignable(a, b)) or (variance == 2 and self.assignable(b, a)))
                if all(matches(a, b, v) for a, b, v in zip(descriptor['arguments'], expected['arguments'], descriptor['variance'])) and len(descriptor['variance']) == len(descriptor['arguments']):
                    return True
            if descriptor: todo.extend(descriptor['interfaces'])
            todo.append(self.parents.get(current))
        return False

    def enumerable_element(self, type_name):
        descriptor = self.meta['types'].get(type_name)
        if descriptor and descriptor['definition'] == 'System.Collections.Generic.IEnumerable`1': return descriptor['arguments'][0]
        return None

    def is_type(self, value, target):
        if isinstance(value, (CliArray, CliString)):
            if target == 'System.Collections.IEnumerable': return True
            wanted = self.enumerable_element(target)
            actual = value.element if isinstance(value, CliArray) else 'System.Char'
            if wanted is not None:
                return actual == wanted or self.default(actual) is None and self.default(wanted) is None and self.assignable(actual, wanted)
        if isinstance(value, CliDelegate): return target in ('System.Object', 'System.Delegate', 'System.MulticastDelegate', value.type)
        if isinstance(value, CliBox): return target in ('System.Object', 'System.ValueType') or self.assignable(value.type, target)
        return super().is_type(value, target)

    def box(self, value, type_name):
        if self.default(type_name) is not None: return CliBox(type_name, self.coerce(value, type_name))
        return value

    def unbox(self, value, type_name, address=False):
        if self.default(type_name) is None and not address: return self.cast(value, type_name)
        self.nonnull(value)
        if not isinstance(value, CliBox) or value.type != type_name:
            self.fail('System.InvalidCastException', 'Specified cast is not valid.')
        if address: return CliRef(lambda: value.value, lambda v: setattr(value, 'value', self.coerce(v, type_name)))
        return self.coerce(value.value, type_name)

    def call(self, method_id, args, virtual=False, constrained=None):
        method = self.meta['methods'][method_id]
        if constrained is not None and method['intrinsic'] == 'object.string':
            return self.string(self.format(self.nonnull(self.dereference(args[0])), constrained))
        if virtual:
            original = self.nonnull(args[0])
            receiver = self.nonnull(self.dereference(original))
            wanted = self.enumerable_element(method['type'])
            if isinstance(receiver, (CliArray, CliString)) and method['name'] == 'GetEnumerator' and (wanted is not None or method['type'] == 'System.Collections.IEnumerable'):
                if isinstance(receiver, CliString):
                    units = receiver.text.encode('utf-16-le', 'surrogatepass')
                    receiver = CliArray('System.Char', [int.from_bytes(units[i:i+2], 'little') for i in range(0, len(units), 2)])
                constructor = self.meta['arrayEnumerators'].get((wanted or receiver.element) + '[]')
                if constructor is None: raise RuntimeError('Array enumeration contract was not linked')
                return self.new_object(constructor, [receiver])
            actual_type = receiver.type if isinstance(receiver, (CliObject, CliBox)) else None
            if method['virtual'] and actual_type:
                method_id = self.meta['types'].get(actual_type, {}).get('vtable', {}).get(method['slot'], method_id)
                method = self.meta['methods'][method_id]
            descriptor = self.meta['types'].get(method['type'])
            if descriptor and descriptor['valueType']:
                if isinstance(receiver, CliBox): args = [self.unbox(receiver, receiver.type, True)] + args[1:]
                elif isinstance(original, CliRef): args = [original] + args[1:]
                else: raise TypeError('Value-type method requires a managed address')
            else: args = [receiver] + args[1:]
        return super().call(method_id, args, False)

    async def await_export(self, name, args, max_steps=100000):
        import asyncio
        if not isinstance(max_steps, int) or max_steps < 1: raise ValueError('max_steps must be a positive integer')
        candidates = [(k, v) for k, v in self.meta['exports'].items() if k == name or k.split('(')[0] == name]
        if len(candidates) != 1: raise ValueError('Use an unambiguous exported method signature')
        result_type = self.meta['methods'][candidates[0][1]]['returns']
        binding = self.meta['asyncBindings'].get(result_type)
        value = self.invoke_export(name, args)
        if binding is None: return value
        for _ in range(max_steps):
            if self.call(binding['completed'], [value]):
                cell = [self.call(binding['getAwaiter'], [value])]
                result = self.call(binding['getResult'], [self.cell_ref(cell, 0, binding['awaiterType'])])
                if isinstance(result, CliString): return result.text
                return bool(result) if binding['resultType'] == 'System.Boolean' else result
            self.call(binding['pump'], [])
            await asyncio.sleep(0)
        raise TimeoutError('Cooperative task exceeded the host pump step budget')

    @staticmethod
    def delegate_equal(a, b):
        if a is b: return True
        if a is None or b is None or a.type != b.type or len(a.invocations) != len(b.invocations): return False
        return all(x[0] is y[0] and x[1] == y[1] for x, y in zip(a.invocations, b.invocations))

    def external(self, method, args):
        op = method['intrinsic']
        if op == 'environment.thread': return 1
        if op == 'delegate.invoke':
            delegate = self.nonnull(args[0]); result = None
            for target, pointer in delegate.invocations:
                callee = self.meta['methods'][pointer]
                receiver = target
                if callee['instance'] and self.meta['types'].get(callee['type'], {}).get('valueType'):
                    receiver = self.unbox(target, callee['type'], True)
                result = self.call(pointer, ([receiver] if callee['instance'] else []) + args[1:])
            return result
        if op in ('delegate.equals', 'delegate.not-equals'):
            result = self.delegate_equal(args[0], args[1])
            return int(not result if op == 'delegate.not-equals' else result)
        if op in ('delegate.combine', 'delegate.remove'):
            a, b = args
            if a is None: return b if op == 'delegate.combine' else None
            if b is None: return a
            if a.type != b.type: self.fail('System.ArgumentException', 'Delegates must be of the same type.')
            if op == 'delegate.combine': return CliDelegate(a.type, a.invocations + b.invocations)
            for start in range(len(a.invocations) - len(b.invocations), -1, -1):
                part = CliDelegate(a.type, a.invocations[start:start + len(b.invocations)])
                if self.delegate_equal(part, b):
                    remaining = a.invocations[:start] + a.invocations[start + len(b.invocations):]
                    return CliDelegate(a.type, remaining) if remaining else None
            return a
        return super().external(method, args)


class CliFunction:
    __slots__ = ('method',)
    def __init__(self, method): self.method = method


class CliDelegate:
    __slots__ = ('type', 'invocations')
    def __init__(self, type_name, invocations): self.type, self.invocations = type_name, tuple(invocations)
